### DISCUSSION TITLE
C++ target performance

### DISCUSSION THREAD

**Discussion by ghost** [antlr/antlr4#3782] [General]
I found the C++ target to have poor performance, when parsing and creating a parse tree.
How does that compare to the Java target?  

When parsing a 227,000-line file of C code the ANTLR parser required 14 seconds. 
A parser generated by LRSTAR required 0.1 seconds.   It looks very strange.

Has anybody compared the C++ performance to the Java performance ?

**Comment by KvanTTT** [comment]
Similar issues:

* [Parsing with C++ target 5x slower than Java](https://github.com/antlr/antlr4/issues/2584)
* [[Cpp] I found Cpp-runtime is 4x slower than Java-runtime](https://github.com/antlr/antlr4/issues/3426)
* [Performance across targets](https://github.com/antlr/antlr4/issues/3111)

**Comment by ghost** [comment]
OK, Java target is 5x faster than C++.  
In my experience LRSTAR parsers are 140x faster than ANTLR C++ targets. 
So then only 28x faster than ANTLR Java targets.  That's still very strange.
Does anybody want to try LRSTAR to verify my findings?  I'm not an ANTLR
or Java guy. 
https://sourceforge.net/project/lrstar

**Comment by kaby76** [comment]
I don't think there's a good Antlr grammar for C. The one in [grammars-v4/c](https://github.com/antlr/grammars-v4/tree/master/c) requires huge lookaheads for expressions and declarations, and is ambiguous for parameterDeclarations. The [expression rules](https://github.com/antlr/grammars-v4/blob/e1f826ecd25000bac20ecc8ce1a074b7db32c460/c/C.g4#L33-L152) are unnatural for Antlr. Similar rules in the [Java Language Spec](https://docs.oracle.com/javase/specs/jls/se11/html/jls-15.html) [were refactored for performance](https://github.com/antlr/grammars-v4/blob/e1f826ecd25000bac20ecc8ce1a074b7db32c460/java/java/JavaParser.g4#L571). It seems as though it's derived from ISO 9899, but I can't tell what refactorings were made.

**Comment by ghost** [comment]
I used the same C grammar for both ANTLR and LRSTAR, which seemed to work fine. 
The grammar was a large subset (90%) of the full specification.  LRSTAR can read ANTLR 
grammars (without Labels and a few other operators).  Maybe it's like comparing apples
to oranges.  I did not study the C++ target code.  It's possible to create C++ code that is 
not fast.  It was about 2 years ago.  I should repeat the test again and take a closer look 
at the C++ target.  Also I should use the Java target for a comparison.  

Unfortunately, this stuff takes time.

**Reply by kaby76** [reply]
I would recommend reading the paper at https://www.antlr.org/papers/LL-star-PLDI11.pdf, generating a parser for a simple grammar and looking over the generated code and ATNs. Antlr performs a powerset construction from ATNs at parse time for branch predictions in the recursive descent parse. There's no simple table representing either DFA or ATN graphs.

**Reply by ghost** [reply]
OK, thanks.  But 14 seconds seems like a lot of time to ser up branch predictions.  Apples to oranges, I guess.  I'll look at the paper.

**Reply by KvanTTT** [reply]
If I understand correctly, the more actual article exists: https://www.antlr.org/papers/allstar-techreport.pdf It's not draft.

**Reply by ghost** [reply]
So, what you are saying is that you have to pay a performance price to avoid the complexity of LALR(k) or LL(k)?
But it was 140x slower than LALR(1).  I suspect that most of the problem is NOT in the ALL(*) algorithm.  It has to 
be somewhere else, serialization, unicode to char, constructors-destructors, or something in the C++ target.  I 
haven't gotten around to redoing the test yet, but it's on my to-do list.

**Reply by kaby76** [reply]
Paul, could you post the 227k-line file so I can do a quick profile?

**Comment by ghost** [comment]
[test.input.txt](https://github.com/antlr/antlr4/files/9112561/test.input.txt)

**Reply by kaby76** [reply]
Paul, Thanks for that file.

Yes, [grammars-v4/c](https://github.com/antlr/grammars-v4/tree/master/c) takes a long time--12s for the C# target. So, I decided to dive into some profiling using the C# target.

Of the 12 seconds to run, 6 seconds is in AdaptivePredict(), and most of that in Closure() or Closure_(). Most of the calls to AdaptivePredict() don't create any new DFA state, but closure()/closure_() is called nonetheless. So, even though we may start in a DFA, most of the time it doesn't help: the parser is constantly going back to an ATN config set to interpret the next state to move to on a symbol because the DFA state is parse stack sensitive (see Function 4 in the paper, `if D0 stack sensitive then input.seek(start); return LLpredict(A, start, γ0);`, section `3.1 Predictions sensitive to the call stack`). I think it may be stack sensitive because the grammar is ambiguous, but I don't know if semantic predicates would help in disambiguating.

Yes, it's spending most of the time predicting what to do in the recursive descent.

**Reply by ghost** [reply]
Wow.  Is that the same behavior with Java target? I assume it is.   
It seems like a heavy price to pay, but ALL(*) seems to work for most people. 
Honey badger?

**Reply by ghost** [reply]
Thanks kaby76.  It would have taken me a long time to figure that out, too long.  
Closure?  Kind of like what a parser generator does, but at parsing time?  Then
that's what makes ALL(*) so powerful.  

My LRSTAR does `LR(*)` at parsing time, but it's shallow `LR(*)`.  It looks ahead
only as far as it needs to, but it refuses to be recursive (call itself), so shallow. 
I would like to make a deep LR(*) that is recursive in the near future, but I'm 
feeling like Terence Parr when he said, ANTLR 4 is the last one. 

Thanks again.
